### ISSUE TITLE
Retrying subscription on failed auth for private channel

### ISSUE THREAD

**Issue by Hypnopompia** [pusher/pusher-js#255]
**Do you want to request a *feature* or report a *bug*?**
I'm not sure if this is a bug or just incomplete documentation.

**What is the current behavior?**
When I subscribe to a private channel, and the auth server fails for some reason, the channel gets stuck in a weird state that I can't retry. Trying pusher.subscribe() on the same channel names returns the existing channel object but doesn't perform an auth request to subscribe to the channel.

**If the current behavior is a bug, please provide the steps to reproduce and
if possible a minimal demo of the problem via https://jsfiddle.net or similar.

```
var pusher = new Pusher('APP_KEY');
var channel = pusher.subscribe('private-channel');
channel.bind('pusher:subscription_error', function(status) {
  if(status == 408 || status == 503){
    // retry?
  }
});
```

this is the sample code from the documentation: https://pusher.com/docs/client_api_guide/client_events#subscription_error

but there's no information on the proper method to retry. calling pusher.subscribe() again doesn't do anything.

**What is the expected behavior?**
pusher.subscribe() on a channel that failed to auth should subscribe the channel.

As a work around, I dug through the source and found the channel has a method subscribe() that retries the auth process. However, in the code this method was marked as internal use only, so I can't count on it being there in the future. It does work to retry though. It just looks like the pusher.subscribe() method doesn't handle the case where a channel is in a failed auth state and retries.

**Comment by hph** [comment]
Hi @Hypnopompia, and thanks for raising the issue. It does appear that `pusher.subscribe` should handle this, but it's fine to use `channel.subscribe` in the meantime.

**Comment by hph** [comment]
I just went back and tested this and you should be able to do something like this (you can create a helper to do this for you if you find yourself doing it more than once):

```javascript
const pusher = new Pusher('APP_KEY');
let channel = pusher.subscribe('private-channel');
channel.bind('pusher:subscription_error', status => {
  if (status == 408 || status == 503) {
    channel = pusher.subscribe('private-channel');
  }
});
```

**Comment by pietro909** [comment]
@Hypnopompia did this answer work for you? I'm experiencing exactly the same issue right now, but the solution proposed by @hph is not working as I get a memory error (call stack exceeded).

I tried using a timeout:

```javascript
const pusher = new Pusher('APP_KEY');
let channel = pusher.subscribe('private-channel');
channel.bind('pusher:subscription_error', status => {
    setTimeout(() => {
        channel = pusher.subscribe('private-channel');
    }, 5000);
});
```

The problem is that after the first iteration, I don't even see `pusher:subscription_error` anymore: it just stops doing anything.

**Comment by pietro909** [comment]
Also, the error I'm coping with is `401` since is caused by an invalid token.

**Comment by liamhubers** [comment]
There is a little bit of a lack of documentation for this, so this is how I did it:
```
new Pusher(process.env.PUSHER_KEY, {
  authorizer: (channel) => ({
    authorize: async (socketId, callback) => {
      try {
        const response = await Api.post('/v2/pusher/auth', {
          socket_id: socketId,
          channel_name: channel.name,
        });

        callback(null, { auth: response.auth });
      } catch (err) {
        callback(true, err.status);
      }
    },
  }),
});
```

**Comment by pesterhazy** [comment]
This issue is still present in the latest release (7.0.0) - could you reopen?

Steps to repro

- use a custom authorizer
- pick a channel, Pusher.subscribe()
- see the custom authorized function be called; let the first authentication fail (easy to do synthetically by returning a callback with status code 418 I'm a Teapot)
- Pusher.unsubscribe()
- Pusher.subscribe()

Expected behavior

- the custom authorizer is called again, giving you a chance to re-try auth

Actual behavior

- the authorized is not called gain
- as described in the description, the channel is now in a weird state which you can't get out of anymore

What seems to work for me is to manually set

```
channel.subscriptionPending = false
```

in the authorizer function in the `catch` clause. However, that's using internal state and not the exposed interface. Is there a better way to work around this Pusher issue? Is a fix forthcoming?

To @liamhubers's point, I don't think using a custom authorizer is a solution, because after a failed attempt, the authorizer function isn't called anymore (because subscriptionPending is true).

**Comment by TomKemp** [comment]
Hi @pesterhazy,

I've been able to replicate the issue. I think the custom authorizer solutions above are suggesting that you only ever return successes to the custom authorizer callback and handle failures/retries in your custom authorizer. That way, the channel never gets into the bad state.

But that isn't ideal so I've started working on a fix that I think should stop the channel getting onto the bad state. I'll let you know how that progresses.
